### PR TITLE
Bridges V2 refactoring backport and `pallet_bridge_messages` simplifications

### PR DESCRIPTION
**Pull request by bkontur**
## Summary

This PR contains migrated code from the Bridges V2 [branch](https://github.com/paritytech/polkadot-sdk/pull/4427) from the old `parity-bridges-common` [repo](https://github.com/paritytech/parity-bridges-common/tree/bridges-v2).
Even though the PR looks large, it does not (or should not) contain any significant changes (also not relevant for audit).
This PR is a requirement for permissionless lanes, as they were implemented on top of these changes.

## TODO

- [x] generate fresh weights for BridgeHubs
- [x] run `polkadot-fellows` bridges zombienet tests with actual runtime 1.2.5. or 1.2.6 to check compatibility
  - :ballot_box_with_check:  working, checked with 1.2.8 fellows BridgeHubs 
- [x] run `polkadot-sdk` bridges zombienet tests
  - :ballot_box_with_check: with old relayer in CI (1.6.5) 
- [x] run `polkadot-sdk` bridges zombienet tests (locally) - with the relayer based on this branch - https://github.com/paritytech/parity-bridges-common/pull/3022
- [x] check/fix relayer companion in bridges repo - https://github.com/paritytech/parity-bridges-common/pull/3022
- [x] extract pruning stuff to separate PR https://github.com/paritytech/polkadot-sdk/pull/4944

Relates to: https://github.com/paritytech/parity-bridges-common/issues/2976
Relates to: https://github.com/paritytech/parity-bridges-common/issues/2451

### CODE REVIEW

**Comment by bkontur** [comment]
bot bench cumulus-bridge-hubs --runtime=bridge-hub-rococo --pallet=pallet_bridge_grandpa

**Comment by bkontur** [comment]
bot bench cumulus-bridge-hubs --runtime=bridge-hub-rococo --pallet=pallet_bridge_parachains
bot bench cumulus-bridge-hubs --runtime=bridge-hub-rococo --pallet=pallet_bridge_messages
bot bench cumulus-bridge-hubs --runtime=bridge-hub-rococo --pallet=pallet_bridge_relayers
bot bench cumulus-bridge-hubs --runtime=bridge-hub-westend --pallet=pallet_bridge_grandpa
bot bench cumulus-bridge-hubs --runtime=bridge-hub-westend --pallet=pallet_bridge_parachains
bot bench cumulus-bridge-hubs --runtime=bridge-hub-westend --pallet=pallet_bridge_messages
bot bench cumulus-bridge-hubs --runtime=bridge-hub-westend --pallet=pallet_bridge_relayers
bot bench cumulus-bridge-hubs --runtime=bridge-hub-rococo --subcommand=xcm --pallet=pallet_xcm_benchmarks::generic
bot bench cumulus-bridge-hubs --runtime=bridge-hub-westend --subcommand=xcm --pallet=pallet_xcm_benchmarks::generic
bot bench cumulus-assets --runtime=asset-hub-rococo --pallet=pallet_xcm_bridge_hub_router
bot bench cumulus-assets --runtime=asset-hub-westend --pallet=pallet_xcm_bridge_hub_router

**Comment by bkontur** [comment]
bot clean
bot bench cumulus-bridge-hubs --runtime=bridge-hub-rococo --pallet=pallet_bridge_grandpa
bot bench cumulus-bridge-hubs --runtime=bridge-hub-rococo --pallet=pallet_bridge_parachains
bot bench cumulus-bridge-hubs --runtime=bridge-hub-rococo --pallet=pallet_bridge_messages
bot bench cumulus-bridge-hubs --runtime=bridge-hub-rococo --pallet=pallet_bridge_relayers
bot bench cumulus-bridge-hubs --runtime=bridge-hub-westend --pallet=pallet_bridge_grandpa
bot bench cumulus-bridge-hubs --runtime=bridge-hub-westend --pallet=pallet_bridge_parachains
bot bench cumulus-bridge-hubs --runtime=bridge-hub-westend --pallet=pallet_bridge_messages
bot bench cumulus-bridge-hubs --runtime=bridge-hub-westend --pallet=pallet_bridge_relayers
bot bench cumulus-bridge-hubs --runtime=bridge-hub-rococo --subcommand=xcm --pallet=pallet_xcm_benchmarks::generic
bot bench cumulus-bridge-hubs --runtime=bridge-hub-westend --subcommand=xcm --pallet=pallet_xcm_benchmarks::generic
bot bench cumulus-assets --runtime=asset-hub-rococo --pallet=pallet_xcm_bridge_hub_router
bot bench cumulus-assets --runtime=asset-hub-westend --pallet=pallet_xcm_bridge_hub_router

**Comment by bkontur** [comment]
bot clean

File: bridges/modules/grandpa/src/lib.rs [hunk 1]
```diff
@@ -1443,11 +1443,14 @@ mod tests {
 	}
 
 	#[test]
-	fn parse_finalized_storage_proof_rejects_proof_on_unknown_header() {
+	fn verify_vec_db_storage_rejects_unknown_header() {
```
[the remark this answers is not in the document]
Comment (bkontur) [reply bridges/modules/grandpa/src/lib.rs]: Fixed here: https://github.com/paritytech/polkadot-sdk/pull/4935/commits/1ed7df78ee122cd4e5cf12ece85109a084f9f475

File: bridges/modules/messages/src/benchmarking.rs [hunk 2]
```diff
@@ -411,51 +494,39 @@ benchmarks_instance_pallet! {
...
-		let i in EXPECTED_DEFAULT_MESSAGE_LENGTH .. EXPECTED_DEFAULT_MESSAGE_LENGTH * 16;
-
-		let relayer_id_on_source = T::bridged_relayer_id();
-		let relayer_id_on_target = account("relayer", 0, SEED);
-		T::endow_account(&relayer_id_on_target);
-
-		// mark messages 1..=20 as delivered
-		receive_messages::<T, I>(20);
-
+	// #[benchmark(extra)]
```
[the remark this answers is not in the document]
Comment (bkontur) [reply bridges/modules/messages/src/benchmarking.rs]: thank you, done here: https://github.com/paritytech/polkadot-sdk/pull/4935/commits/1ed7df78ee122cd4e5cf12ece85109a084f9f475

**Review by acatangiu — APPROVED** [review]
Code changes look good!

Should do follow-up investigation on grandpa submit proof weight for bridge-hub-rococo.

**Comment by bkontur** [comment]
bot clean
bot bench cumulus-bridge-hubs --runtime=bridge-hub-rococo --pallet=pallet_bridge_grandpa
bot bench cumulus-bridge-hubs --runtime=bridge-hub-westend --pallet=pallet_bridge_grandpa

File: bridges/modules/messages/src/weights.rs [hunk 3]
```diff
@@ -17,9 +17,9 @@
 //! Autogenerated weights for pallet_bridge_messages
 //!
 //! THIS FILE WAS AUTO-GENERATED USING THE SUBSTRATE BENCHMARK CLI VERSION 4.0.0-dev
-//! DATE: 2023-03-23, STEPS: `50`, REPEAT: `20`, LOW RANGE: `[]`, HIGH RANGE: `[]`
+//! DATE: 2023-06-22, STEPS: `50`, REPEAT: `20`, LOW RANGE: `[]`, HIGH RANGE: `[]`
```
[the remark this answers is not in the document]
Comment (bkontur) [reply bridges/modules/messages/src/weights.rs:20]: https://github.com/paritytech/parity-bridges-common/issues/2393

**Pull request merged**